### PR TITLE
Add precision configuration option

### PR DESCRIPTION
**Pull request by SimonKocurek**
Currently it is possible for user/attacker to send `max * 2` requests if they send half of the requests just before `hits` map gets cleared... to mitigate this issue, requests are split into buckets (effectively dividing whole `windowMs` into multiple sub-windows).

Inspired by https://www.figma.com/blog/an-alternative-approach-to-rate-limiting

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -120,6 +120,14 @@ Note: with non-default stores, you may need to configure this value twice, once
 
 Defaults to `60000` (1 minute).
 
+### precision
+
+Number of request buckets inside time window. Buckets group requests based on timestamp for efficient cleanup.
+
+With precision set to `1` all remembered requests are reset, when time window expires. Increasing this value too much will result in decreased performance.
+
+Defaults to `4`.
```
Comment (nfriedly) [inline README.md:129]: This should clarify that `precision` only applies to the built-in memory store. 

(Other stores generally track the time per-user and so essentially have "infinity precision".)

File: package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "express-rate-limit",
-  "version": "5.3.0",
+  "version": "5.4.0",
```
Comment (nfriedly) [inline package.json:3]: You should leave this out of the PR. After it's merged in, I'll use the `npm version minor` command which both increments the version number and makes a git tag.

File: lib/express-rate-limit.js [hunk 3]
```diff
@@ -10,6 +10,7 @@ function RateLimit(options) {
       statusCode: 429, // 429 status = Too Many Requests (RFC 6585)
       headers: true, //Send custom rate limit header with limit and remaining
       draft_polli_ratelimit_headers: false, //Support for the new RateLimit standardization headers
+      precision: 4, //Number of request buckets inside time window. Buckets serve as a performant alternative to tracking timestamp of each request and clearing them one by one.
```
Comment (nfriedly) [inline lib/express-rate-limit.js:13]: Lets make the default `1`, otherwise this could end up surprising some people. I might bump it up to 4 in the next major release, though.

File: lib/memory-store.js [hunk 4]
```diff
@@ -6,41 +6,60 @@ function calculateNextResetTime(windowMs) {
   return d;
 }
```
Comment (nfriedly) [inline lib/memory-store.js:7]: `calculateNextResetTime` should also take the precision into account. It might need to look through the buckets too (?) We basically want it to give out the next time that the current user could make a request.

**Review by nfriedly — CHANGES_REQUESTED** [review]
Hey, thanks for doing this and sorry for taking so long to review it.

I left a few requests below, but overall I really like this.

File: lib/memory-store.js (hunk 4 above)
Comment (SimonKocurek) [reply lib/memory-store.js:7]: There is iterator which takes precision into account and moves onto next bucket when this time expires:
```js
  const bucketTime = windowMs / precision;
  const bucketResetInterval = setInterval(this.resetBucket, bucketTime);
```

`NextResetTime` is just pure `windowMs`, since it is sent via HTTP headers to inform the client when will server forget his requests. That happens after all buckets are cleared.

I might have wrong understanding of this header though. If you are not convinced I can change it to also take `precision` into account. :)

Comment (SimonKocurek) [reply lib/memory-store.js:7]: After reading [https://tools.ietf.org/id/draft-polli-ratelimit-headers-00.html#rfc.section.3.3](https://tools.ietf.org/id/draft-polli-ratelimit-headers-00.html#rfc.section.3.3) I see that I've understood it wrongly.  Thought it is used in header with time when user will be 100% safe to send new request.

File: test/memory-store-test.js [hunk 5]
```diff
@@ -4,7 +4,7 @@ const MemoryStore = require("../lib/memory-store.js");
 
 describe("MemoryStore store", () => {
   it("sets the value to 1 on first incr", (done) => {
-    const store = new MemoryStore(-1);
+    const store = new MemoryStore(-1, 4);
```
Comment (nfriedly) [inline test/memory-store-test.js:7]: Lets undo these changes and instead add some tests that specifically check different precision values.

```suggestion
    const store = new MemoryStore(-1);
```

File: test/memory-store-test.js [hunk 6]
```diff
@@ -21,7 +21,7 @@ describe("MemoryStore store", () => {
   });
 
   it("increments the key for the store each incr", (done) => {
-    const store = new MemoryStore(-1);
+    const store = new MemoryStore(-1, 4);
```
Comment (nfriedly) [inline test/memory-store-test.js:24]:
```suggestion
    const store = new MemoryStore(-1);
```

File: test/memory-store-test.js [hunk 7]
```diff
@@ -40,7 +40,7 @@ describe("MemoryStore store", () => {
   });
 
   it("resets the key for the store when used with resetKey", (done) => {
-    const store = new MemoryStore(-1);
+    const store = new MemoryStore(-1, 4);
```
Comment (nfriedly) [inline test/memory-store-test.js:43]:
```suggestion
    const store = new MemoryStore(-1);
```

File: test/memory-store-test.js [hunk 8]
```diff
@@ -61,7 +61,7 @@ describe("MemoryStore store", () => {
   });
 
   it("resets all keys for the store when used with resetAll", (done) => {
-    const store = new MemoryStore(-1);
+    const store = new MemoryStore(-1, 4);
```
Comment (nfriedly) [inline test/memory-store-test.js:64]:
```suggestion
    const store = new MemoryStore(-1);
```

File: test/memory-store-test.js [hunk 9]
```diff
@@ -94,7 +94,7 @@ describe("MemoryStore store", () => {
   });
 
   it("resets all keys for the store when the timeout is reached", (done) => {
-    const store = new MemoryStore(50);
+    const store = new MemoryStore(50, 4);
```
Comment (nfriedly) [inline test/memory-store-test.js:97]:
```suggestion
    const store = new MemoryStore(-1);
```

File: test/memory-store-test.js [hunk 10]
```diff
@@ -147,7 +147,7 @@ describe("MemoryStore store", () => {
     });
 
     it("can run in electron where setInterval does not return a Timeout object with an unset function", (done) => {
-      const store = new MemoryStore(-1);
+      const store = new MemoryStore(-1, 4);
```
Comment (nfriedly) [inline test/memory-store-test.js:150]:
```suggestion
    const store = new MemoryStore(-1);
```

File: test/memory-store-test.js [hunk 11]
```diff
@@ -171,7 +171,7 @@ describe("MemoryStore store", () => {
   });
 
   it("decrements the key for the store each decrement", (done) => {
-    const store = new MemoryStore(-1);
+    const store = new MemoryStore(-1, 4);
```
Comment (nfriedly) [inline test/memory-store-test.js:174]:
```suggestion
    const store = new MemoryStore(-1);
```

**Review by nfriedly — CHANGES_REQUESTED** [review]
Sorry again for taking so long.

I left a few minor suggestions below, and then the last thing I'd like to see are a few tests exercising different values of precision to check how many hits are remaining after different amounts of time. You can model it on this one or bring a clock mocking library if you prefer: 
https://github.com/nfriedly/express-rate-limit/blob/c4f9c46ddf6d6f86861007df431d775deb5f379c/test/memory-store-test.js#L96-L133

**Comment by gamemaker1** [comment]
Hi @SimonKocurek,

Thanks for taking the time to make this contribution!

Do you wish to continue working on this PR? I see that this PR has had no activity since September last year. In the time between, we have rewritten the library in Typescript and made quite a few changes to the `Store` interface too. I completely understand if you do not have the time to continue work on this, in which case I could maybe implement this myself (@nfriedly would you still be interested in seeing this merged?).

Thanks!

**Comment by SimonKocurek** [comment]
@gamemaker1 I have no plans in the near future to finish this. Feel free to handle this PR. (AFAIK nobody has requested this feature, so even closing is fine by me.)

**Comment by gamemaker1** [comment]
> AFAIK nobody has requested this feature

There are a couple of issues that would be solved with this increased precision feature like https://github.com/nfriedly/express-rate-limit/issues/212 and https://github.com/nfriedly/express-rate-limit/issues/207. I think this would be a good enhancement to the memory store, while still keeping it simple.

> I have no plans in the near future to finish this.

I'll close this PR for now, and start a new one from the `master` branch, to avoid dealing with the merge conflicts. Thanks for creating this PR and giving us this awesome idea :D

**Comment by nfriedly** [comment]
Yeah, I'm ambivalent on this one. I don't think it's a bad idea, but my main issue is that the mechanics of the MemoryStore are already complex enough that they often confuse people, and this just dials it up a notch. Perhaps it would make more sense in an external store?
